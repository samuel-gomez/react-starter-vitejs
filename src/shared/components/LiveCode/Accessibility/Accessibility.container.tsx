/* eslint-disable react/no-unused-prop-types */
import Accessibility from './Accessibility';
import { useAxe } from './Accessibility.hook';

type TAccessibilityContainer = {
  hideComponent?: boolean;
  title?: string;
  icon?: string;
  code?: string;
  ariaLabel?: string;
  useAxeFn?: typeof useAxe;
};

const AccessibilityContainer = ({ ariaLabel, code, useAxeFn = useAxe }: TAccessibilityContainer) => {
  const { errors, results } = useAxeFn({ html: `[aria-label="${ariaLabel}"]`, code });

  return <Accessibility errors={errors} results={results} />;
};

export default AccessibilityContainer;
