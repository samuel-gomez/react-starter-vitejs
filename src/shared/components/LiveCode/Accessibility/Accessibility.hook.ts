import axe, { AxeResults } from 'axe-core';
import { useEffect, useState } from 'react';

export type TResults = {
  errors?: Error;
  results?: AxeResults;
};
export type TsetResult = ({ errors, results }: TResults) => void;

export const axeRunAccessibility = (html: string, setResults: TsetResult) => axe.run(html, {}, (errors, results) => setResults({ errors, results }));

export type TuseAxe = {
  html: string;
  code?: string;
  axeRunAccessibilityFn?: typeof axeRunAccessibility;
};

export const useAxe = ({ html, code, axeRunAccessibilityFn = axeRunAccessibility }: TuseAxe) => {
  const [results, setResults] = useState<TResults>();

  useEffect(() => {
    axeRunAccessibilityFn(html, setResults);
  }, [axeRunAccessibilityFn, html, code]);

  return { errors: results?.errors, results: results?.results };
};
