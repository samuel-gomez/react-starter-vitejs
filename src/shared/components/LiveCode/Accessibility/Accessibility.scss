@import '@axa-fr/react-toolkit-core/src/common/scss/core.scss';

$colors: (
  1: $color-red-error,
  2: $color-green-4,
  3: $color-orange-dark,
);

%custom-radius {
  padding: 0;
  margin: 1px;
  background-color: $color-white;
  border: none;
}

%font-weight {
  .af-tabs__link {
    font-weight: 500;
  }
}

.af-tabs {
  &--accessibility-content {
    .af-tabs__pane {
      padding: 1.25rem;
    }
    .af-tabs__content {
      border: none;
      box-shadow: none;
    }
    @each $id, $color in $colors {
      .af-tabs__item--custom-radius:nth-child(#{$id}) {
        .af-tabs__link {
          color: $color;
        }
      }
    }
    .af-tabs__item--custom-radius {
      @extend %custom-radius;
      @extend %font-weight;
      border-radius: 0;
      margin-top: 0;
      height: auto;
    }
    .af-tabs__item--custom-radius.af-tabs__item--active {
      @extend %custom-radius;
      @extend %font-weight;
      border-bottom: solid 3px $color-axa;
    }
  }
}
