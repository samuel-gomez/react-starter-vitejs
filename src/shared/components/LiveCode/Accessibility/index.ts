export { default } from './Accessibility.container';
