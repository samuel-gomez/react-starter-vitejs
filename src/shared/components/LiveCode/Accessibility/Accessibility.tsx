import { Tabs } from '@axa-fr/react-toolkit-all';
import { type TResults } from './Accessibility.hook';
import './Accessibility.scss';
import AccordionResults from './AccordionResults';
import EmptyResultsAccessiblity from './EmptyResultsAccessiblity';
import TitleTab from './TitleTab';

const Accessibility = ({ results }: TResults) => (
  <Tabs classModifier="accessibility-content">
    <Tabs.Tab title={<TitleTab results={results?.violations} label="Violations" icon="alert" />} classModifier="has-icon-left custom-radius">
      {results && results.violations.length > 0 ? (
        <AccordionResults results={results.violations} />
      ) : (
        <EmptyResultsAccessiblity message="Pas de violations" />
      )}
    </Tabs.Tab>
    <Tabs.Tab title={<TitleTab results={results?.passes} label="Passes" icon="ok" />} classModifier="has-icon-left custom-radius">
      {results && results.passes.length > 0 ? <AccordionResults results={results.passes} /> : <EmptyResultsAccessiblity message="Pas de passes" />}
    </Tabs.Tab>
    <Tabs.Tab title={<TitleTab results={results?.incomplete} label="Incomplete" icon="retweet" />} classModifier="has-icon-left custom-radius">
      {results && results.incomplete.length > 0 ? (
        <AccordionResults results={results.incomplete} />
      ) : (
        <EmptyResultsAccessiblity message="Pas d'incomplete" />
      )}
    </Tabs.Tab>
  </Tabs>
);
export default Accessibility;
