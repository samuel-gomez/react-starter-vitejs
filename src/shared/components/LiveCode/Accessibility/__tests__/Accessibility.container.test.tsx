import { render } from '@testing-library/react';
import AccessibilityContainer from '../Accessibility.container';
import { useAxe } from '../Accessibility.hook';

const ariaLabelMock = '';

const returnUseAxe = {
  errors: undefined,
  results: {
    violations: [
      {
        description: 'test',
        help: 'test',
        helpUrl: 'url',
        impact: 'critical',
        tags: ['tag1'],
        id: 'id-test',
      },
    ],
    passes: [],
    incomplete: [],
  },
} as unknown as ReturnType<typeof useAxe>;

const useAxeFnMock = () => returnUseAxe;

describe('<AccessibilityContainer />', () => {
  it('Render <AccessibilityContainer /> with ariaLabel and useAxeFn', () => {
    const { asFragment } = render(<AccessibilityContainer ariaLabel={ariaLabelMock} useAxeFn={useAxeFnMock} />);

    expect(asFragment()).toMatchSnapshot();
  });
  it('Render <AccessibilityContainer /> with ariaLabel', () => {
    const { asFragment } = render(<AccessibilityContainer ariaLabel={ariaLabelMock} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
