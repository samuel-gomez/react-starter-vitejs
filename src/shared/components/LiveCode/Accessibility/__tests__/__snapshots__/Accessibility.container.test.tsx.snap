// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<AccessibilityContainer /> > Render <AccessibilityContainer /> with ariaLabel 1`] = `
<DocumentFragment>
  <div
    class="af-tabs af-tabs--accessibility-content"
  >
    <ul
      class="af-tabs__control"
    >
      <li
        class="af-tabs__item af-tabs__item--has-icon-left af-tabs__item--custom-radius af-tabs__item--active"
      >
        <button
          class="af-tabs__link"
          id="0"
          type="button"
        >
          <i
            class="glyphicon glyphicon-alert"
          />
           (0) Violations
        </button>
      </li>
      <li
        class="af-tabs__item af-tabs__item--has-icon-left af-tabs__item--custom-radius"
      >
        <button
          class="af-tabs__link"
          id="1"
          type="button"
        >
          <i
            class="glyphicon glyphicon-ok"
          />
           (0) Passes
        </button>
      </li>
      <li
        class="af-tabs__item af-tabs__item--has-icon-left af-tabs__item--custom-radius"
      >
        <button
          class="af-tabs__link"
          id="2"
          type="button"
        >
          <i
            class="glyphicon glyphicon-retweet"
          />
           (0) Incomplete
        </button>
      </li>
    </ul>
    <div
      class="af-tabs__content"
    >
      <div
        class="af-tabs__pane"
      >
        <div
          class="empty-accessiblity"
        >
          Pas de violations
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<AccessibilityContainer /> > Render <AccessibilityContainer /> with ariaLabel and useAxeFn 1`] = `
<DocumentFragment>
  <div
    class="af-tabs af-tabs--accessibility-content"
  >
    <ul
      class="af-tabs__control"
    >
      <li
        class="af-tabs__item af-tabs__item--has-icon-left af-tabs__item--custom-radius af-tabs__item--active"
      >
        <button
          class="af-tabs__link"
          id="0"
          type="button"
        >
          <i
            class="glyphicon glyphicon-alert"
          />
           (1) Violations
        </button>
      </li>
      <li
        class="af-tabs__item af-tabs__item--has-icon-left af-tabs__item--custom-radius"
      >
        <button
          class="af-tabs__link"
          id="1"
          type="button"
        >
          <i
            class="glyphicon glyphicon-ok"
          />
           (0) Passes
        </button>
      </li>
      <li
        class="af-tabs__item af-tabs__item--has-icon-left af-tabs__item--custom-radius"
      >
        <button
          class="af-tabs__link"
          id="2"
          type="button"
        >
          <i
            class="glyphicon glyphicon-retweet"
          />
           (0) Incomplete
        </button>
      </li>
    </ul>
    <div
      class="af-tabs__content"
    >
      <div
        class="af-tabs__pane"
      >
        <div
          aria-multiselectable="true"
          class="af-accordion af-accordion--light af-accordion--accessibility"
          role="tablist"
        >
          <div
            class="af-accordion__item"
          >
            <div
              class="af-accordion__item-header"
              id="id-test"
              role="tab"
            >
              <a
                aria-controls="id-testbody"
                aria-expanded="false"
                class="af-accordion__item-toggle"
                data-toggle="isOpen"
                href="#id-testbody"
              >
                <h3
                  class="af-accordion__item-title"
                >
                  <span
                    class="glyphicon glyphicon-menu-down"
                  />
                  test 
                  <span
                    class="af-badge af-badge--error"
                  >
                    critical
                  </span>
                </h3>
              </a>
            </div>
            <div
              aria-labelledby="id-test"
              class="af-accordion__collapse af-accordion__collapse--close"
              id="id-testbody"
              role="tabpanel"
              style="max-height: 0;"
            >
              <div
                class="af-accordion__block"
              >
                <div
                  class="af-accordion--accessibility-wrapper-body"
                >
                  <p>
                    test
                  </p>
                  <p>
                    <a
                      href="url"
                      rel="noreferrer"
                      target="_blank"
                    >
                      Plus d'infos...
                    </a>
                  </p>
                </div>
                <div
                  class="af-accordion--accessibility-wrapper-body"
                >
                  <span>
                    Tags : 
                  </span>
                  <span
                    class="af-badge af-badge--info"
                  >
                    tag1
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
